### ISSUE TITLE
Switch exhaustiveness error with enum and 'case null' 

### ISSUE THREAD

**Issue by rovarga** [eclipse-jdt/eclipse.jdt.core#2582]
For this bit of code:
```
    public enum Foo {
        FOO, BAR;
    }

    public void dispatch(final Foo foo) {
        switch (foo)  {
            // forces JEP-441 exhaustiveness
            case null -> throw new NullPointerException("null foo");
            case FOO -> foo();
            case BAR -> bar();
        }
    }

    private void foo() {
    }

    private void bar() {
    }
```

javac-21.0.3 happily accepts it and does the right thing.
JDT elicits an `An enhanced switch statement should be exhaustive; a default label expected` error.

Seems to be present it current development (I tested with JDT 3.19.600.v20240616-1800).

**Comment by srikanth-sankaran** [comment]
> Seems to be present it current development (I tested with JDT 3.19.600.v20240616-1800).

Are you sure about the versions ? This seems to be a duplicate of https://github.com/eclipse-jdt/eclipse.jdt.core/issues/277 - the test case is almost identical and as of the comment https://github.com/eclipse-jdt/eclipse.jdt.core/issues/277#issuecomment-1851795520 dated Dec 12, 2023 itself the problem is not reproducible. I will release a regression test proving that it works and close this.

I am using this test case: 

```
enum Foo {
        FOO, BAR;
    }

public class X {
    public void dispatch(final Foo foo) {
        switch (foo)  {
            // forces JEP-441 exhaustiveness
            case null -> throw new NullPointerException("null foo");
            case FOO -> foo();
            case BAR -> bar();
        }
    }

    private void foo() {
    	System.out.println("Foo");
    }

    private void bar() {
    	System.out.println("Bar");
    }
    public static void main(String[] args) {
    	new X().dispatch(Foo.FOO);
    	new X().dispatch(Foo.BAR);
    	try {
    	new X().dispatch(null);
    	} catch (NullPointerException npe) {
    		System.out.println("Null");
    	}
	}
}
```

In future always provide full test cases - the methods `dispatch` `foo` and `bar` are not enclosed in any class - Or are you saying you are using the preview feature implicitly named classes ?? I doubt that since the code is lacking `main` method and won't compile without it and you claim it compiles.

**Comment by srikanth-sankaran** [comment]
I have added a regression test to prove this works ok - and that test passes. If you still see the problem, please open with a full test case and details needed to reproduce against next I build

**Comment by stephan-herrmann** [comment]
@srikanth-sankaran I haven't yet caught up with the bells and whistles of new switches, but I wonder: for old switches we suggest to add `default:` *even if all enum constants are covered*, to safeguard against evolution of the enum. Is there something like this regarding new switches, too?

**Comment by srikanth-sankaran** [comment]
> @srikanth-sankaran I haven't yet caught up with the bells and whistles of new switches, but I wonder: for old switches we suggest to add `default:` _even if all enum constants are covered_, to safeguard against evolution of the enum. Is there something like this regarding new switches, too?

From https://bugs.eclipse.org/bugs/show_bug.cgi?id=388314:

```
The cases of a switch expression must be exhaustive; for all possible values there must be a matching switch label. (Obviously switch statements are not required to be exhaustive.)

In practice this normally means that a default clause is required; however, in the case of an enum switch expression that covers all known constants, a default clause is inserted by the compiler to indicate that the enum definition has changed between compile-time and runtime. Relying on this implicit default clause insertion makes for more robust code; now when code is recompiled, the compiler checks that all cases are explicitly handled. **Had the developer inserted an explicit default clause (as is the case today) a possible error will have been hidden.**
```

Compiler will insert code to throw a MatchException will be thrown when no case matches due to the switch ceasing to be exhaustive due to partial recompile.

**Comment by stephan-herrmann** [comment]
Thanks, I think we'll have to live with different concepts of "completeness", and users will have to make a non-trivial decision anyway:
* when a new enum constant appears, do they like seeing a MatchException or would they prefer to handle this in a default branch up-front? (enum could come from a library, which is perhaps upgraded at deployment time).

Our previous (complex) discussion in this field was https://bugs.eclipse.org/bugs/show_bug.cgi?id=374605

I just checked that our rich set of warning options works for both kinds of switches alike:
```java
package test;

enum NUM { ONE, TWO, THREE; }

public class Switches {
	void old(NUM n) {
		switch (n) { // The switch over the enum type NUM should have a default case
			case ONE, TWO, THREE:
				System.out.println("known");
		}
		switch (n) {
			case ONE, TWO, THREE:
				System.out.println("known");
			default:
				System.out.println("safety");
		}
		switch (n) { // The enum constant THREE should have a corresponding case label in this enum switch on NUM. To suppress this problem, add a comment //$CASES-OMITTED$ on the line above the 'default:'
			case ONE, TWO:
				System.out.println("known");
				break;
			default:
				System.out.println("other");
		}
		switch (n) {
			case ONE, TWO:
				System.out.println("known");
				break;
			//$CASES-OMITTED$
			default:
				System.out.println("other");
		}
	}
	void newStyle(NUM n) {
		switch (n) {  // The switch over the enum type NUM should have a default case
			case ONE, TWO, THREE ->
				System.out.println("known");
		}
		switch (n) {
			case ONE, TWO, THREE ->
				System.out.println("known");
			default ->
				System.out.println("safety");
		}
		switch (n) { // The enum constant THREE should have a corresponding case label in this enum switch on NUM. To suppress this problem, add a comment $CASES-OMITTED$ on the line above the 'default:'
			case ONE, TWO -> 
				System.out.println("known");
			default ->
				System.out.println("other");
		}
		switch (n) {
			case ONE, TWO -> 
				System.out.println("known");
			//$CASES-OMITTED$
			default ->
				System.out.println("other");
		}
	}
}
```
Warnings are shown as they a raised with these options:
![image](https://github.com/eclipse-jdt/eclipse.jdt.core/assets/24587985/5873ee2d-cbf7-4a06-b724-0b512165bc57)

Tiny glitch: `default:` in the last warning message should either omit `:` or say `default ->` :)

Anyway, none of this is about exhaustiveness as per JLS, just convenience warnings.

**Comment by rovarga** [comment]
@srikanth-sankaran I may have the versions confused, it is the first time I am delving into how my IDE works :) I am working on code changes that heavily use sealed interface (and ran into the NPEs you fixed). I am definitely still seeing the error, but after upgrading to the nightly build, but the NPEs are gone. So weird, but something happened :) I'll try to delve deeper into Eclipse next week, but that's moot -- thanks for adding the test case!

**Comment by rovarga** [comment]
@stephan-herrmann right, that's exactly what I found -- and I think the completeness lies in the trick the code I posted does.

If you'd write your last switch as:
```
		switch (n) {
			case ONE, TWO -> 
				System.out.println("known");
			case null -> throw NullPointerException();
		}
```

you would get a compile-time error if you introduce NUM.FOUR.

**Comment by stephan-herrmann** [comment]
Combining the trick with `case null` with enabling all optional warnings pointed right to the culprit.

Over to #2611

**Comment by srikanth-sankaran** [comment]
Both of you speak of a trick  - I am a little bit unclear, what "trick" is being talked about here ??!

**Comment by srikanth-sankaran** [comment]
> Thanks, I think we'll have to live with different concepts of "completeness", and users will have to make a non-trivial decision anyway:
> 
> * when a new enum constant appears, do they like seeing a MatchException or would they prefer to handle this in a default branch up-front? (enum could come from a library, which is perhaps upgraded at deployment time).
> 
> Our previous (complex) discussion in this field was https://bugs.eclipse.org/bugs/show_bug.cgi?id=374605
> 
> I just checked that our rich set of warning options works for both kinds of switches alike:

Very good. Appreciate that you are ever watchful of code changes that intersect you past contributions :)

I'll add this test case in the upcoming PR.

> Tiny glitch: `default:` in the last warning message should either omit `:` or say `default ->` :)

Can address this too.

It turns out in `org.eclipse.jdt.internal.compiler.ast.SwitchStatement.resolve(BlockScope)` at the line that says: `// check default case for all kinds of switch:` we haven't even determined exhaustiveness of enum switches :-( which happens only below.

**Comment by srikanth-sankaran** [comment]
I have put together an interim fix at https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2613 - this entire file is going to see a rewrite/cleanup in the coming weeks.

**Comment by srikanth-sankaran** [comment]
> I am working on code changes that heavily use sealed interface

In that case you may want to know that several defects in the area of sealed types are being closed for 4.33 M1. Thanks for upgrading when it becomes available and reporting any problems.
